packages/apollo-server-defer/README.md

```diff
@@ -0,0 +1,173 @@
+- Try to maximize code reuse by exporting types and utility functions from graphql.js, making a PR if necessary.
+- Restrict peer dependency on graphql.js to versions that we have tested with.
+- @skip and @include should take priority over @defer
+
+#### Apollo Client:
+
+- No changes required
+- Initial implementation of @defer support should come as a Apollo Link. Reads from a socket connection or some other event stream, keeps the partial response in memory, merging patches as they come and pushing it through the link stack.
+
+#### Errors:
+
+This refers specifically to the errors that occur when streaming deferred fields. Errors on the initial response will be handled as per normal.
+
+- Resolver level errors can be sent along with the patch. These will be merged with the errors array to create the full response.
+- Network errors (i.e when the websocket connection is broken). Ideally, a network error should not nullify the results that have already been sent succesfully. However, the client should expose the network status to indicate whether all the patches have been sent successfully.  
+
+#### Transport:
+
+- @defer requires a transport layer that is able to stream or push data to the client, like websockets or server side events. If the transport layer does not support this, Apollo Server should fallback to normal execution and ignore @defer.
+- Note: Deferred queries are not able to be refetched in isolation. So links like apollo-link-retry and apollo-link-error might need to ignore patches for now.
+
+#### Restrictions on @defer usage:
+
+- _Mutations:_ already execute serially, so there does not seem to be any use case for @defer. We should throw an error if @defer is applied to the root mutation type. However, @defer should apply normally for the selection set on a mutation.
+- _Non-Nullable Types_: Deferring non-nullable types may lead to unexpected behavior when errors occur, since errors will propagate up to a nullable field. We do not want a deferred field to propagate errors up to its parent. The fields returned with the initial response should not be nulled by a deferred error - it makes things easier to reason about as all patches are strictly additive.
+
+- _@defer should apply regardless of data availability_
+
+  - Even if the deferred fields are available in memory immediately, it should not be sent with the initial response.
+
+    ```// For example, even if the entire "asset" is queried from
+    // the database as a single object, we still defer sending
+    // the "reviews" field
+    query {
+      asset {
+        id
+        title
+        reviews @defer
+      }
+    }
+    ```
+
+  - The reason that this behavior is useful is because some fields can incur high bandwidth to transfer, slowing down initial load.
+  - We could allow the user to control this behavior, by taking an optional waitFor argument:
```

Comment:
I think this is really interesting and we probably also want some kind of timeout so a socket doesn't get stuck for some reason. @peggyrayzis this could have interesting supsense tie ins

Reply:
Agreed! It would be so cool to sync the same `waitFor` times throughout the presentation and data loading layer!